components/places/android/src/main/java/org/mozilla/places/PlacesConnection.kt

```diff
@@ -17,40 +17,89 @@ import java.util.concurrent.atomic.AtomicLong
 import java.util.concurrent.atomic.AtomicReference
 
 /**
- * An implementation of a [PlacesAPI] backed by a Rust Places library.
+ * An implementation of a [PlacesApi] backed by a Rust Places library.
  *
  * @param path an absolute path to a file that will be used for the internal database.
  * @param encryption_key an optional key used for encrypting/decrypting data stored in the internal
  *  database. If omitted, data will be stored in plaintext.
  */
-class PlacesConnection(path: String, encryption_key: String? = null) : PlacesAPI, AutoCloseable {
+class PlacesApi(path: String, encryption_key: String? = null) : PlacesApiInterface, AutoCloseable {
     private var handle: AtomicLong = AtomicLong(0)
-    private var interruptHandle: InterruptHandle
 
     init {
-        try {
-            handle.set(rustCall { error ->
-                LibPlacesFFI.INSTANCE.places_connection_new(path, encryption_key, error)
-            })
-        } catch (e: InternalPanic) {
-
-            // Places Rust library does not yet support schema migrations; as a very temporary quick
-            // fix to avoid crashes of our upstream consumers, let's delete the database file
-            // entirely and try again.
-            // FIXME https://github.com/mozilla/application-services/issues/438
-            if (e.message != "sorry, no upgrades yet - delete your db!") {
-                throw e
+        handle.set(rustCall { error ->
+            LibPlacesFFI.INSTANCE.places_api_new(path, encryption_key, error)
+        })
+    }
+
+    companion object {
+        // These numbers come from `places::db::ConnectionType`
+        private const val READ_ONLY: Int = 1
+        private const val READ_WRITE: Int = 2
+    }
+
+    override fun open_reader(): ReadablePlacesConnection {
+        val conn_handle = rustCall { error ->
+            LibPlacesFFI.INSTANCE.places_connection_new(handle.get(), READ_ONLY, error)
+        }
+        return ReadablePlacesConnection(conn_handle);
+    }
+
+    override fun open_writer(): WritablePlacesConnection {
+        val conn_handle = rustCall { error ->
+            LibPlacesFFI.INSTANCE.places_connection_new(handle.get(), READ_WRITE, error)
+        }
+        return WritablePlacesConnection(conn_handle);
+    }
+
+    @Synchronized
+    override fun close() {
```

Comment:
I think it might be kind of unintuitive that closing the PlacesAPI doesn't close all the connections. It also can be used to 'defeat' our system. Unless I'm mistaken, someone can get two writer connections as follows:

- Open one api
- Get it's writer connection (W1)
- Close api
- Open another api to the same DB
- Get it's writer connection (W2)
- We now have W1 and W2.

It doesn't seem that implausible that something could accidentally do this either. We have the ability to prevent this, and so we probably should.